backend/app/spread/actions.py

```diff
@@ -30,25 +31,36 @@ def update_balances_with_spread(
     if token_to_divest_from == spread_token_symbol:
         return resized_balances
 
-    zeroes_series = make_zeroes(start, end)
+    def make_spread_token_unit_series():
+        return hist_spread_price.loc[start:end] / hist_spread_price.loc[start]
+
+    divest_token_hist_balances = balances.usd_balances[token_to_divest_from]
 
     try:
-        start_divest_balance = balances.balances[token_to_divest_from].loc[start]
+        divest_token_start_usd_balance = divest_token_hist_balances.loc[start]
     except KeyError:
-        spread_additional_token_balance: pd.Series = zeroes_series
-    else:
-        spread_additional_token_balance: pd.Series = (
-            zeroes_series + start_divest_balance * spread_percentage / 100.0
-        )
-        resized_balances.balances[token_to_divest_from] *= (
+        divest_token_start_usd_balance = 0
+
+    # Do not resize divest token balance if no swap was possible.
+    if divest_token_start_usd_balance > 0:
+        resized_balances.usd_balances[token_to_divest_from] *= (
             100 - spread_percentage
         ) / 100.0
 
-    if spread_token_symbol in resized_balances.balances:
-        spread_additional_token_balance = spread_additional_token_balance.add(
-            resized_balances.balances[spread_token_symbol], fill_value=0
+    spread_token_additional_usd_balance = (
+        divest_token_start_usd_balance
+        * spread_percentage
+        / 100
+        * make_spread_token_unit_series()
```

Comment:
Similar to my comment above. I get the sense that this unit series is used to elegantly get the USD value of the treasury's "spread token" balance post-spread action, but it doesn't immediately become clear to me.

Honestly, this implementation is very elegant, but maybe this function overall could be broken down to multiple functions at a later refractor and maybe the variable themselves could be named better for more clarity. Some type hints for these variables would also be helpful.

Reply:
I'll try refactoring a bit.

Reply:
I hope these explanations help.

Reply:
Yup, that does it for me! With that I think this PR is ready to merge.